java-frontend/src/test/java/org/sonar/java/UCFGJavaVisitorTest.java

```diff
@@ -86,38 +86,269 @@ public void visit_java_file() {
+            .assignTo(aux0, call("A#foo()LA;").withArgs(Expression.THIS), new LocationInFile(FILE_KEY, 4,11,4,16))
+            .assignTo(aux1, call("A#foo()LA;").withArgs(aux0), new LocationInFile(FILE_KEY, 4,11,4,22))
+            .assignTo(aux2,
+                call("__id").withArgs(new Expression.FieldAccess(aux1, variableWithId("field"))),
+                new LocationInFile(FILE_KEY, 4,11,4,28))
+            .assignTo(aux3, call("A#foo()LA;").withArgs(aux2), new LocationInFile(FILE_KEY, 4,11,4,34))
+            .ret(aux3, new LocationInFile(FILE_KEY, 4,4,4,35)))
+        .build();
+    assertCodeToUCfg("class A {\n" +
+        "  A foo() { A a = new A(); a.field = new A(); return a; } \n" +
+        "  A bar() {\n" +
+        "    return foo().foo().field.foo();\n" +
+        "  }\n" +
+        "  A field;\n" +
+        "}\n", expectedUCFG);
+  }
+
+  @Test
+  public void field_access_on_this() {
+    Expression.Variable x = variableWithId("x");
+    Expression.Variable aux0 = variableWithId("%0");
+    Expression.Variable aux1 = variableWithId("%1");
+    Expression.Variable aux2 = variableWithId("%2");
+    Expression.Variable aux3 = variableWithId("%3");
+    Expression.Variable aux4 = variableWithId("%4");
+    UCFG expectedUCFG = UCFGBuilder.createUCFGForMethod("A#method()Ljava/lang/String;")
+      .addBasicBlock(newBasicBlock("1")
+        .assignTo(aux0,
+          call("__id").withArgs(new Expression.FieldAccess(variableWithId("field"))),
+          new LocationInFile(FILE_KEY, 5,15,5,25))
+        .assignTo(x, call("__id").withArgs(aux0), new LocationInFile(FILE_KEY, 5,4,5,26))
+        .assignTo(aux1,
+          call("__id").withArgs(new Expression.FieldAccess(variableWithId("field"))),
+          new LocationInFile(FILE_KEY, 6,11,6,16))
+        .assignTo(aux2,
+          call("__id").withArgs(new Expression.FieldAccess(new Expression.ClassName("A"), variableWithId("FIELD"))),
+          new LocationInFile(FILE_KEY, 6,19,6,24))
+        .assignTo(aux3, call("__concat").withArgs(aux1, aux2), new LocationInFile(FILE_KEY, 6,11,6,24))
+        .assignTo(aux4, call("__concat").withArgs(aux3, x), new LocationInFile(FILE_KEY, 6,11,6,28))
+        .ret(aux4, new LocationInFile(FILE_KEY, 6,4,6,29)))
+      .build();
+    assertCodeToUCfg("class A {\n" +
+        "  String field;\n" +
+        "  static String FIELD;\n" +
+        "  String method() {\n" +
+        "    String x = this.field;\n" +
+        "    return field + FIELD + x;\n" +
+        "  }\n" +
+        "}\n", expectedUCFG);
+  }
+
+  @Test
+  public void field_access_on_this_from_inner_class() {
+    Expression.Variable aux0 = variableWithId("%0");
+    Expression.Variable aux1 = variableWithId("%1");
+    UCFG expectedUCFG = UCFGBuilder.createUCFGForMethod("A$B#innerFoo(Ljava/lang/String;)Ljava/lang/String;")
+        .addBasicBlock(newBasicBlock("1")
+            .assignTo(aux0,
+                // FIXME what do we expect here?
```

Comment:
I think we are gonna live with this for the time being. This is quite a corner case.